### PR TITLE
Implement phase 2 weighted target selection

### PR DESCRIPTION
**Pull request by toto-dev**
I strongly suggest to review the single commit rather then the whole change

### CODE REVIEW

File: ctools/defragment_sharded_collection.py [hunk 1]
```diff
@@ -696,116 +687,127 @@ async def get_remain_chunk_imbalance(center, target_chunk):
...
             # due to parallelism the chunks might now be on the same shard            
 
             left_chunk = chunks_max_index.get(pickle.dumps(c['min']))
             right_chunk = chunks_min_index.get(pickle.dumps(c['max']))
             
-            # Exclude overweight target shards
-            if (left_chunk is not None and right_chunk is not None) and (left_chunk['shard'] != right_chunk['shard']):
-                if total_shard_size[left_chunk['shard']] > total_shard_size[right_chunk['shard']] * args.shard_imbalance_frac:
+            if left_chunk and right_chunk:
+                for dest_chunk in [left_chunk, right_chunk]:
```
Comment (ppolato) [inline ctools/defragment_sharded_collection.py:701]: I would add some comment along the voting logic to explain the rationale behind each step

File: ctools/defragment_sharded_collection.py [hunk 2]
```diff
@@ -696,116 +687,127 @@ async def get_remain_chunk_imbalance(center, target_chunk):
...
+                
+                if right_chunk['votes'] > left_chunk['votes']:
                     left_chunk = None
-                elif total_shard_size[right_chunk['shard']] > total_shard_size[left_chunk['shard']] * args.shard_imbalance_frac:
-                    right_chunk = None
-                else:
-                    pass
+
 
             if left_chunk is not None:
```
Comment (ppolato) [inline ctools/defragment_sharded_collection.py:728]: I believe that we could easily collapse in a single function the logic to move and merge the center with the chosen candidate (once you order the chunks by increasing minKey, the code for merging and updating the in-memory data structures becomes the same).

File: ctools/defragment_sharded_collection.py [hunk 3]
```diff
@@ -821,7 +823,7 @@ async def phase_2():
 
         with tqdm(total=total_chunks_to_process, unit=' chunks') as progress:
             iteration = 0
-            while iteration < 25:
+            while True:
```
Comment (ppolato) [inline ctools/defragment_sharded_collection.py:826]: I slightly prefer having a variable here telling me the condition for leaving the loop - by this way, it becomes easier to understand where and why we are done.

**Review by ppolato — CHANGES_REQUESTED** [review]

File: ctools/defragment_sharded_collection.py [hunk 4]
```diff
@@ -696,116 +687,127 @@ async def get_remain_chunk_imbalance(center, target_chunk):
                     continue
                 else:
                     break
+                
+            if total_shard_size[shard] * args.shard_imbalance_frac < max_shard_size and not center_size_kb == 0:
```
[the remark this answers is not in the document]
Comment (toto-dev) [reply ctools/defragment_sharded_collection.py:691]: The imbalance threshold goes from `1` to `1.5` ,so this formula tells that the current shard should donate (only if its size times the threshold is less the the size of the shard with the bigger amount of data) or (if the chhunk he is donating has no data (size=0))

File: ctools/defragment_sharded_collection.py (hunk 1 above)
Comment (toto-dev) [reply ctools/defragment_sharded_collection.py:701]: Yes I'll do

File: ctools/defragment_sharded_collection.py (hunk 2 above)
Comment (toto-dev) [reply ctools/defragment_sharded_collection.py:728]: I tried it and it is not exactly the same so I decided to leave it for another moment.

File: ctools/defragment_sharded_collection.py (hunk 3 above)
Comment (toto-dev) [reply ctools/defragment_sharded_collection.py:826]: Unfortunately python doesn't have the do/while construct, so I would need to introduce somenthing like:
```
while not in first_round and condition_still_meet:
   loop
```

File: ctools/defragment_sharded_collection.py (hunk 2 above)
Comment (ppolato) [reply ctools/defragment_sharded_collection.py:728]: How about something like this? (assumed to be a nested function)

```
    def move_chunk_and_merge_with_sibling(chunk, sibling, is_chunk_before_sibling):
        source_shard = chunk['shard']
        target_shard = sibling['shard']
        sibling_size = await get_chunk_size(sibling)
        new_size = sibling_size + center_size_kb
        first_merge_member = chunk if is_chunk_before_sibling else sibling
        second_merge_member = sibling if is_chunk_before_sibling else chunk
        merge_bounds = [first_merge_member['min'], second_merge_member['max']]
        if not args.dryrun:
            if source_shard != target_shard:
                await coll.move_chunk(chunk, target_shard)
            await coll.merge_chunks([first_merge_member, second_merge_member])
            if args.write_chunk_size:
                await coll.try_write_chunk_size(merge_bounds, target_shard, new_size)
        else:
            progress.write(f'Moving chunk [{chunk["min"]},{chunk["max"]}] from {shard} to {target_shard}, '
                            f'merging {merge_bounds}, new size: {fmt_kb(new_size)}')
        # update local map, 
        chunks_id_index.pop(second_merge_member['_id']) # only first chunk is kept
        chunks_min_index.pop(pickle.dumps(second_merge_member['min']))
        chunks_max_index.pop(pickle.dumps(first_merge_member['max']))
        chunks_max_index[pickle.dumps(second_merge_member['max'])] = first_merge_member
        first_merge_member['shard'] = target_shard
        first_merge_member['merged'] = True
        first_merge_member['max'] = c['max']
        first_merge_member['defrag_collection_est_size'] = new_size

        if source_shard != target_shard:
            total_shard_size[source_shard] -= center_size_kb
            total_shard_size[target_shard] += center_size_kb
            max_shard_size = max(max_shard_size, total_shard_size[target_shard]) 
            total_moved_data_kb += center_size_kb

        # update stats for merged chunk (source)
        progress.update(1)
        #update stats for merged chunk (destination)
        if sibling_size <= small_chunk_size_kb and new_size > small_chunk_size_kb:
            progress.update(1)

        await exec_throttle()
        begin_time = time.monotonic()
```

File: ctools/defragment_sharded_collection.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ppolato) [reply ctools/defragment_sharded_collection.py:691]: This is fine to me, but I believe that we should then modify the description of the parameter.

**Pull request closed without merging**